File: application/src/java/wandledi/example/pages/HomePage.java
package wandledi.example.pages;

import java.util.Collection;
import java.util.LinkedList;
import wandledi.core.Attribute;
import wandledi.core.Scroll;
import wandledi.example.models.BlogEntry;
import wandledi.example.models.Comment;
import wandledi.java.html.PageImpl;
import wandledi.java.html.Plan;
import wandledi.java.html.Selectable;
import wandledi.java.html.SelectableElement;

import static wandledi.java.Switchboard.linkToUri;
import static wandledi.java.Switchboard.linkToId;

/**
 *
 * @author Markus Kahl
 */
public class HomePage extends PageImpl {

    private boolean login = false;

    public void beforeAction(String msg, String homeLink) {

        get(new Attribute("rel", "stylesheet")).setAttribute("href",
                linkToUri("/css/main.css"));
        get(".homelink").setAttribute("href", homeLink);
        if (msg != null) {
            get("#right").insert(msg);
        }
        if (login) {
            get("a", "href", "login").setAttribute("href", "post");
            get("button").replace(true, "Post Entry");
        }
    }

    public void index(Collection<BlogEntry> entries) {

        if (entries.isEmpty()) {
            get(".entry").setAttribute("id", "last");
            get(".heading").replace(true, "No Entries yet");
            get(".text").replace(true, "There are no entries yet. Why don't you start writing one?");
            get(".footer").hide();
        } else {
            get(".entry").foreachIn(entries).apply(new Plan<BlogEntry>() {
                public void execute(SelectableElement element, BlogEntry entry) {
                    if (last()) {
                        element.setAttribute("id", "last");
                    }
                    produceEntry(element, entry);
                }
            });
        }
    }

    public void login() {

        setFile("/home/index.xhtml");
        includeInEntry("/home/login.xhtml", this);
    }

    public void post() {

        post(null, null, null, null);
    }

    public void post(String msg, String author, String title, String content) {

        Scroll scroll = new Scroll("Post");
        if (author != null) {
            scroll.get(new Attribute("name", "author")).setAttribute("value", author);
        }
        if (title != null) {
            scroll.get(new Attribute("name", "title")).setAttribute("value", title);
        }
        if (content != null) {
            scroll.get(new Attribute("name", "content")).replace(true, content);
        }
        setFile("/home/index.xhtml");
        includeInEntry("/home/post.xhtml", scroll, this);
        get(".entry").changeAttribute("class", "form $val");
    }

    public void comments(final BlogEntry entry) {

        Collection<Comment> comments = new LinkedList<Comment>();

        comments.add(new Comment()); // turn first one into commented blog entry
        comments.addAll(entry.getComments());
        comments.add(new Comment()); // include comment form here
        setFile("/home/index.xhtml");
        get(".entry").foreachIn(comments).apply(new Plan<Comment>() {
            public void execute(SelectableElement element, Comment comment) {
                if (first()) { // blog entry
                    produceEntry(element, entry);
                } else {
                    if (last()) { // comment form
                        element.setAttribute("id", "last");
                        element.changeAttribute("class", "form $val");
                        includeInEntry("/home/comment.xhtml", element);
                    } else { // comment
                        produceComment(element, comment);
                        if (login) {
                            includeDeleteButton(element, comment, entry.getId());
                        }
                    }
                    element.changeAttribute("class", "comment $val");
                }
            }
        });
    }

    private void produceComment(SelectableElement element, Comment comment) {

        element.get(".heading").replace(true, "Comment from " + comment.getAuthor());
        element.get(".subheading").replace(true, comment.getDate().toString());
        element.get(".text").replace(true, comment.getContent());
        element.get(".footer").hide();
    }

    public void comment(BlogEntry entry) {

        comment(null, entry, null, null, null);
    }

    public void comment(String msg, BlogEntry entry, String author, String email, String content) {

        Scroll scroll = new Scroll("Post");
        if (author != null) {
            scroll.get(new Attribute("name", "author")).setAttribute("value", author);
        }
        if (email != null) {
            scroll.get(new Attribute("name", "email")).setAttribute("value", email);
        }
        if (content != null) {
            scroll.get(new Attribute("name", "content")).replace(true, content);
        }
        setFile("/home/index.xhtml");
        includeInEntry("/home/comment.xhtml", scroll, this);
        get(".entry").changeAttribute("class", "form $val");
        if (msg != null) {
            get("#right").insert(msg);
        }
    }

    private void produceEntry(SelectableElement element, BlogEntry entry) {

        element.get(".heading").replace(true, entry.getTitle());
        element.get(".user").replace(true, entry.getAuthor());
        element.get(".date").replace(true, entry.getDate().toString());
        element.get(".text").replace(true, entry.getContent());
        element.get(".footer").insert(entry.getComments().size() + " ");
        element.get(new Attribute("href", "comments")).setAttribute("href",
                linkToId("home", "comments", entry.getId()));
    }

    private void includeDeleteButton(SelectableElement element, Comment comment, long bid) {

        Scroll scroll = new Scroll();

        scroll.get("a").setAttribute("href",
                linkToId("home", "deleteComment", comment.getId()) + "?bid=" + bid);
        element.get("br").includeFile("/home/delete.xhtml", scroll);
    }

    private void includeInEntry(String file, Scroll scroll, Selectable selectable) {

        if (scroll == null) {
            selectable.get("p").includeFile(file);
        } else {
            selectable.get("p").includeFile(file, scroll);
        }
        selectable.get("span").hide();
        selectable.get("br").hide();
    }

    private void includeInEntry(String file, Selectable selectable) {

        includeInEntry(file, null, selectable);
    }

    public boolean isLogin() {
        return login;
    }

    public void setLogin(boolean login) {
        this.login = login;
    }
}

File: scala-app/src/main/scala/wandledi/scala/example/pages/HomePage.scala
package wandledi.scala.example.pages

import wandledi.scala.example.models.BlogEntry
import wandledi.scala.example.models.Comment
import wandledi.scala.html.SelectableElement
import wandledi.scala.html._
import wandledi.java.Switchboard.linkToUri
import wandledi.core.Attribute
import wandledi.core.Scroll
import wandledi.java.Switchboard.linkToId
import wandledi.scala.Implicits._
import collection.JavaConversions._

class HomePage extends Page {

  var loggedIn: Boolean = false

  def beforeAction(msg: Option[String], homeLink: String) {	  
    get("rel" -> "stylesheet").setAttribute("href", linkToUri("/css/main.css"))
    get(".homelink").setAttribute("href", homeLink)
    get("#right").insert(msg.get) unless msg.isEmpty
    if (loggedIn) {
      get("href" -> "login").setAttribute("href", "post")
      get("button").replace(true, "Post Entry")
    }
  }

  def index(entries: Seq[BlogEntry]) {
    if (entries.isEmpty) {
      get(".entry").setAttribute("id", "last")
      get(".heading").replace(true, "No Entries yet")
      get(".text").replace(true, "There are no entries yet. Why don't you start writing one?")
      get(".footer").hide()
    } else {
      get(".entry").foreachWithIndexIn(entries) { (e, entry, index) =>
        e.setAttribute("id", "last") provided index == entries.size - 1
        produceEntry(e, entry)
      }
    }
  }

  def login() {
    setFile("/home/index.xhtml");
    includeInEntry("/home/login.xhtml", this)((scroll)=>{})
  }
  
  def post(msg: Option[String], author: Option[String], title: Option[String], content: Option[String]) {
    setFile("/home/index.xhtml")
    includeInEntry("/home/post.xhtml", this) { page =>
      page.get("name" -> "author").setAttribute("value", author.get) unless author.isEmpty
      page.get("name" -> "title").setAttribute("value", title.get) unless title.isEmpty
      page.get("name" -> "content").replace(true, content.get) unless content.isEmpty
    }
    get(".entry").changeAttribute("class")("form " + _)
  }

  def post() {
    post(None, None, None, None)
  }

  def comments(entry: BlogEntry) {
    val comments = (new Comment :: entry.getComments.toList) :+ new Comment
    setFile("/home/index.xhtml")
    get(".entry").foreachWithIndexIn(comments) { (e, comment, index) =>
      if (index == 0) { // blog entry
        produceEntry(e, entry)
        e.get(".footer").hide()
      } else {
        if (index == comments.size - 1) { // comment form
          e.setAttribute("id", "last")
          e.changeAttribute("class")("form " + _)
          includeInEntry("/home/comment.xhtml", e) { page => /* include unchanged */ }
        } else { // comment
          produceComment(e, comment)
          includeDeleteButton(e, comment, entry.getId.longValue) provided loggedIn
          e.changeAttribute("class")("comment " + _)
        }
      }
    }
  }

  def comment(msg: Option[String], entry: BlogEntry,
              author: Option[String], email: Option[String], content: Option[String]) {
    setFile("/home/index.xhtml")
    includeInEntry("/home/comment.xhtml", this) { page =>
      page.get("name" -> "author").setAttribute("value", author.get) unless author.isEmpty
      page.get("name" -> "email").setAttribute("value", email.get) unless email.isEmpty
      page.get("name" -> "content").replace(true, content.get) unless content.isEmpty
    }
    get(".entry").changeAttribute("class")("form " + _)
    get("#right").insert(msg.get) unless msg.isEmpty
  }

  def comment(entry: BlogEntry) {
    comment(None, entry, None, None, None)
  }

  def message_=(msg: String): Unit = get("#right").insert(msg)

  /**Always returns null.
   */
  def message: String = null

  private def includeDeleteButton(e: SelectableElement, comment: Comment, bid: Long) {
    e.get("br").includeFile("/home/delete.xhtml") { page =>
      page.get("a").setAttribute("href", linkToId("home", "deleteComment", comment.id) + "?bid=" + bid)
    }
  }

  private def includeInEntry(file: String, selectable: ScalaSelectable)(magic: (ScalaSelectable) => Unit) {
    selectable.get("p").includeFile(file)(magic)
    selectable.get("span").hide()
    selectable.get("br").hide()
  }

  private def produceEntry(e: SelectableElement, entry: BlogEntry) {
    e.get(".heading").replace(true, entry.title)
    e.get(".user").replace(true, entry.author)
    e.get(".date").replace(true, if (entry.date != null) entry.date.toString else "???")
    e.get(".text").replace(true, entry.content)
    e.get("href" -> "comments").setAttribute("href", linkToId("home", "comments", entry.id))
    if (entry.comments.size > 0) {
      e.get(".footer").insert(entry.comments.size + " ")
    } else {
      e.get("href" -> "comments").replace(true, "add comment")
    }
  }

  private def produceComment(e: SelectableElement, comment: Comment) {
    e.get(".heading").replace(true, "Comment from " + comment.author)
    e.get(".subheading").replace(true, comment.date.toString)
    e.get(".text").replace(true, comment.content)
    e.get(".footer").hide()
  }
}
